compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_folding.mlir

```diff
@@ -0,0 +1,133 @@
+      // CHECK: %[[OK0:.+]], %[[QUERY0:.+]] = hal.device.query<%[[DEVICE]] : !hal.device> key("sys" :: "foo")
+      %ok0, %query0 = hal.device.query<%device : !hal.device> key("sys" :: "foo") : i1, i32
+      // CHECK: cf.cond_br %[[OK0]], ^[[BB_OK0:[a-z0-9]+]], ^[[BB_CONT:[a-z0-9]+]](%[[DUMMY]], %[[DUMMY]] : i32, i32)
+      cf.cond_br %ok0, ^bb_ok0, ^bb_fail
+    // CHECK: ^[[BB_OK0]]:
+    ^bb_ok0:
+      // CHECK: %[[OK1:.+]], %[[QUERY1:.+]] = hal.device.query<%[[DEVICE]] : !hal.device> key("sys" :: "bar")
+      %ok1, %query1 = hal.device.query<%device : !hal.device> key("sys" :: "bar") : i1, i32
+      // CHECK: cf.cond_br %[[OK1]], ^[[BB_CONT]](%[[QUERY0]], %[[QUERY1]] : i32, i32), ^[[BB_CONT]](%[[DUMMY]], %[[DUMMY]] : i32, i32)
+      cf.cond_br %ok1, ^bb_ok1, ^bb_fail
+    // Note that these blocks fold but we want them here so that we test the
+    // multiple return -> single return path.
+    ^bb_ok1:
+      hal.return %query0, %query1 : i32, i32
+    ^bb_fail:
+      hal.return %dummy, %dummy : i32, i32
+
+    // Second block gets spliced in here:
+    // CHECK: ^[[BB_CONT]](%[[BB_QUERY0:.+]]: i32, %[[BB_QUERY1:.+]]: i32):
+      // CHECK: %[[OK2:.+]], %[[QUERY2:.+]] = hal.device.query<%[[DEVICE]] : !hal.device> key("sys" :: "baz")
+      // CHECK: hal.return %[[BB_QUERY0]], %[[BB_QUERY1]], %[[QUERY2]]
+    }
+    // CHECK-NOT: hal.executable.constant.block
+    hal.executable.constant.block(%device: !hal.device) -> i32 as "baz" {
+      %ok2, %query2 = hal.device.query<%device : !hal.device> key("sys" :: "baz") : i1, i32
+      hal.return %query2 : i32
+    }
+  }
+}
+
+// -----
+
+// Tests that unused %device args are dropped. This isn't structurally critical
+// but does make the IR easier to read after propagation may elide device
+// queries.
+
+// CHECK-LABEL: @unused_device_arg
+hal.executable @unused_device_arg {
+  hal.executable.variant @backend, target = #hal.executable.target<"backend", "format"> {
+    // CHECK: hal.executable.constant.block() -> i32 as "foo"
+    hal.executable.constant.block(%device: !hal.device) -> i32 as "foo" {
+      %c0 = arith.constant 0 : i32
+      hal.return %c0 : i32
+    }
+  }
+}
+
+// -----
+
+// Tests that keys get deduplicated in blocks such that only one value is
+// produced. Duplication can happen after merging.
+
+// CHECK-LABEL: @duplicate_keys
+hal.executable @duplicate_keys {
+  hal.executable.variant @backend, target = #hal.executable.target<"backend", "format"> {
+    // CHECK: hal.executable.constant.block() -> (i32, i32) as ("foo", "bar")
+    hal.executable.constant.block() -> (i32, i32, i32) as ("foo", "bar", "foo") {
+      // CHECK-DAG: %[[C0:.+]] = arith.constant 0
+      %c0 = arith.constant 0 : i32
+      // CHECK-DAG: %[[C1:.+]] = arith.constant 1
+      %c1 = arith.constant 1 : i32
+      // CHECK-NOT: arith.constant 2
+      %c2 = arith.constant 2 : i32
+      // CHECK: hal.return %[[C0]], %[[C1]] : i32, i32
+      hal.return %c0, %c1, %c2 : i32, i32, i32
```

Comment:
Since there are two values for `foo` here (`0` and `2`), how is only one selected? Should this be an error instead? Can you give an example of how IR like this would be produced (e.g. with some sample device queries for Vulkan)?

Is the idea that the values are queried at runtime via some mechanism like `vkGetPhysicalDeviceProperties`, so the values are expected to be consistent for a given key? (or if they vary over multiple queries, that's undefined / hopefully eventually consistent behavior anyways?)